### ISSUE TITLE
401 Unauthorized Exception, Docker

### ISSUE THREAD

**Issue by oodoloo** [sealbro/dotnet.garmin.connect#35]
Hello,
I am using this package in a .net 7.0 worker service, which I host in a Docker container. Basically it works, but unfortunately now I always sporadically get an exception "Request [POST: https://connect.garmin.com/modern/di-oauth/exchange] return code 401 (Unauthorized)" and the Docker container stops. First it works yes and then suddenly not anymore. Sometimes 4-5 requests in 10-30 minute intervals work and sometimes only 1-2.
Does anyone possibly have a tip on what it could be or is the problem known? Would like to use the library, but I know slowly unfortunately no more further.

**Comment by oodoloo** [comment]
So I think I was able to figure out a problem now. In the GarminConnectContext class, the MakeHttpRequest method queries on HttpStatusCode.Forbidden. I think it should rather query on HttpStatusCode.Unauthorized to force a ReLogin. Whereas I was wondering why not Refresh of the token is triggered? Does the Garmin interface offer that at all?

Because in the meantime I had implemented it a little differently and since I was also sporadically prompted by garmin to reset my password, do not know if it had to do with too many logins. Although these were then successful, but maybe there is still a security mechanism.
I will test it in any case now times with the said adjustment and see how it behaves.

**Comment by PieterBoeren** [comment]
I also had the issue with the password resets, I solved it by just run the docker container via a cron job (and terminating after a run), instead of running it as a service (on a schedule inside the container itself). Since then Garmin never prompted a password reset.

If you ever figure out a better solution, could you post it here?

**Comment by oodoloo** [comment]
Ok, that's a bit questionable why the approach works, though, or can you explain it?
I do not really understand where the difference is there !??
I had it implemented differently for me before, so that I performed a new login for each query.
I have a WorkerService that periodically retrieves data. Initially I had used a short periodic cycle, though I suspect that quickly resulted in a password reset being forced.
Now I have changed the service so that after each fetch a new random data fetch is set between 10 and 30 minutes. In addition, I now only perform a login when I receive an error with the error code 401, which is why I also performed the above-mentioned error correction.
Currently, the service has been running for 1.5 days - even with a successful re-login - let's see how it goes.

**Comment by sealbro** [comment]
Hi, guys!
If you have the solution to your problem I would be glad to review a PR
Could you add some samples with this problem for a fast start?

I will try to research how to help you

**Comment by oodoloo** [comment]
I have now created a PR for my customization. Since then my service has been running under Docker without receiving any password change request from Garmin so far. But before that, it would sometimes run for a few days without any problems, until I would get a password reset prompt again.
So the way the problem presents itself, it's also difficult to provide examples.
As mentioned before, in my first attempt for using this library, I probably still used it incorrectly and therefore implemented my logic incorrectly. Then the next problem was that (apparently) whenever the token from Garmin expired, my service stopped with an exception. I have now been able to solve this error for the time being through the customization, for which I have now also created a PR.
Now I have to wait and see if the problem with the password reset occurs again.

But I still have one open question:
As I understand it, a new login is performed in the library when the token has expired. Is there not the possibility to perform a refresh of the token?
That would probably improve the occurrence of the password reset problem again and make it less likely to occur, since the logins would be reduced again.
Or is there no possibility of a token refresh at all?
